Enable concretization caching (redux)

johnwparent opened the pull request:
Third time is the charm?

Redux of https://github.com/spack/spack/pull/49589 which was reverted by https://github.com/spack/spack/pull/50851 due to race conditions, which efforts have since been made to remove.

#48198 Introduces the concretization cache, and associated manifest. However it only added testing for the cache itself, not the manifest or cleanup procedures.

This PR:

- [x] Adds tests for cleanup
- [x] Refactors the cache to store base32 hashes like the rest of Spack's hashes
- [x] gzips the cache entries for a smaller memory footprint
- [x] Removes redundant stat calls 
- [x] Less aggressive/error prone cleanup phase
- [x] Rewrite pruning to use LRU instead of FIFO

`FileCache` class was refactored into an abstract base `Cache` class which provides the general cache read/write transaction interface, and a `FileCache` and `DirectoryCache` child classes, `FileCache` which should be functionally identical to the `FileCache` class pre refactor, deals with caches at the individual file level, and `DirectoryCache`, which is mean to provide a directory level or "bucket" caching level.

johnwparent committed:
General improvements to the concretization cache

Add test for cache cleanup
Cleanup existing tests
use correct cache limit config name
Switch misc_cache -> user_cache_path
Remove dropped race condition handling
Improve lockfile cleanup

Document config options

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Concretization cache tests: add lockfile cleanup unit test

A legitimate skip test on Windows

Signed-off-by: John Parent <john.parent@kitware.com>

tgamblin committed:
refactor setup to return a ProblemInstanceBuilder instead of string

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
Simplify newline handling in ProblemIntanceBuilder

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

johnwparent commented on lib/spack/spack/test/concretization/core.py:
```diff
@@ -3229,57 +3230,11 @@ def test_commit_variant_can_be_reused(installed_commit, incoming_commit, reusabl
-    # object
-    for _ in range(5):
-        assert h == spack.concretize.concretize_one("hdf5")
+def extract_cache_metadata():
```
This is rebase cruft.

johnwparent replied:
Removed.

johnwparent commented on lib/spack/spack/solver/asp.py:
```diff
@@ -581,160 +569,124 @@ class ConcretizationCache:
+                    removed = True
+            if removed:
+                try:
+                    self._fc.purge_lock(cache_bucket)
```
I don't love this flow, suggestions for improvement encouraged.

johnwparent replied:
Flow Improved.

tgamblin committed:
concretization cache: strip comments and empty lines from cached input representation

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

johnwparent committed:
ConcretizationCache: Switch to range based locking

Remove bucket based caching approach, all entries are now in flat
concretization cache directory

Removes bucket style locking, all cache entries are locked via
range based locking

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Concretization Cache: add test for migration from uncompressed to compressed entries

Previous Versions of the concretization cache were uncompressed. We now compress them to save space.
Explicitly test that the new concretization cache can handle both compressed and uncompressed entries.

Concretizer cache test: add encoding

Refactor concretization cache cleanup

Remove redundant stat call

Make cleanup LRU

Remove lock purge race condition

Proper loop exit condition when performing cleanup

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent commented on lib/spack/docs/config_yaml.rst:
```diff
@@ -293,7 +293,7 @@ When ``false`` or omitted, all concretization requests will be performed from sc
 Currently this only supports paths on the local filesystem.
 
-Default location is under the :ref:`Misc Cache` at: ``$misc_cache/concretization``
+Default location is under the :ref:`Misc Cache` at: ``$user_cache_path/concretization``
```
Is `Misc Cache` still correct?

tgamblin replied:
I'd rather keep it in the misc_cache -- I think it's just simpler. We have too many configuration paths already.

johnwparent commented on lib/spack/spack/solver/asp.py:
```diff
@@ -607,227 +606,186 @@ class ConcretizationCache:
+        for entry in self.cache_entries():
+            # short timeout, if we can't access it, its either being cleaned
+            # or being used, either way, we don't need to bother here
+            with self.read_transaction(entry, timeout=10) as f:
```
This is a really short timeout (write transaction beneath has the same). Didn't want to kill performance waiting for conc cache cleanup locks, am in no way strongly opinionated about this, very open to suggestions.

tgamblin replied:
10s is a long timeout. See below.

johnwparent committed:
ConcretizationCache: remove generator typing

Either mypy, or pylance, or the rhel8 audits complain, so just get rid of it

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Make intermediate dirs

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
forgot unlink doesn't have missing_ok on 3.7

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent commented on lib/spack/spack/test/data/config/concretizer.yaml:
```diff
@@ -5,3 +5,5 @@ concretizer:
     host_compatible: false
   duplicates:
     strategy: minimal
+  concretization_cache:
```
Is this the right place to set this default for the tests? Some of the test failures looked like they were trying to use the conc cache...

tgamblin committed:
Add typing to parts of asp.py

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
concretization cache: ensure mock fixture is not modifying real config

- [x] Use spack.config.override() to make a temporary scope.

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
concretization cache: ensure that entire Result is preserved by round trip

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
concretizer: remove unused test argument on `Result.to_dict`

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
concretization cache: clean up solve code, break out methods

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
concretization cache: move concretizer cache to the misc cache

I don't think this merits its own cache. It's fine for it to live along with other
package centric metadata in the misc cache.

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
asp: simplify concretizer arg names in locking routines

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin commented on lib/spack/spack/schema/concretizer.py:
```diff
@@ -96,6 +96,15 @@
+                "type": "object",
+                "properties": {
+                    "enable": {"type": "boolean"},
+                    "url": {"type": "string"},
```
I think we can leave this as undocumented or just remove it, depending on how @kwryankrattiger wants to handle it.  It think it is better for this stuff to go in the `misc_cache` by default.

johnwparent replied:
I'll leave this up to @kwryankrattiger, I have no opinion.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -607,228 +643,198 @@ class ConcretizationCache:
+
+        entry_count, bytes_count = 0, 0
+        rm_reg: Dict[pathlib.Path, Tuple[float, int]] = {}
+        for entry in self.cache_entries():
```
This is slow and I worry about it for a number of reasons:

1. Locking can be slow but you read every element in the cache every time.
2. `stat()` can be slow but you `stat` every element in the cache every time.
3. Cleanup only removes the last couple cached entries, so it's going to be expensive on nearly every concretization once the cache fills up. On a slow filesystem this is going to be bad.

I do not think we need both the count size limit and the size limit. The size limit requires that you `stat()` everything, so I think we should throw it out. Also, I'm pretty sure nobody will hit the size limit given the defaults you have here. `3e8` bytes is like 300MB, but compressed specs are maybe ~35k from glancing at them.  1000 specs is going to be like 35 MB, so all that work is likely for nothing.

My suggestion is this:

1. Throw out the size limit.
2. Keep the entry limit, but don't remove *just* enough to get back under the limit -- remove half the cache when you hit the limit. Advantages:
    1. This gives the cache some time to fill up again before you touch 1,000 files -- basically every time you clean up it'll need to crawl back to 1,000 to do something expensive again.
    2. You don't have to lock or stat anything -- you can just do, e.g., `file_count = sum(1 for item in root.iterdir() if item.is_file())` to see if you *need* to clean up.

This will cause the cache to be cleaned much less frequently. Also, if we want to guarantee that we store the last 1,000 things, we can set the limit to 2,000.

johnwparent replied:
Previously you had been interested in maintaining this in an LRU fashion, which would still require stat calls, just double checking that we no longer care about this being LRU.

tgamblin replied:
We do care about it being LRU. But the stat calls are only going to happen when the cache fills up, then not for a while.  So basically:

1. 1000 new concretizations, each lists the directory (cheap), and any intervening hits call utime on their file (cheap)
2. 1001st concretization lists the directory and finds it's too large
3. NOW do a bunch of stats to figure out what was last modified
4. Remove half of the files.
5. 500 more concretizations execute cheaply
6. repeat from 3

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -607,228 +643,198 @@ class ConcretizationCache:
+                # get time of last use so we can implement an LRU removal
+                access_time = entry_stat_info.st_atime
+                mod_time = entry_stat_info.st_mtime
+                most_recent_op_time = access_time if access_time > mod_time else mod_time
```
note that atime isn't reliable.

johnwparent replied:
I actually did not know this, learned something today. Thanks

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -860,10 +868,32 @@ def fetch(self, problem: str) -> Union[Tuple[Result, List], Tuple[None, None]]:
-            if f:
-                result = self._results_from_cache(f)
-                statistics = self._stats_from_cache(f)
+        with self.read_transaction(cache_path) as exists:
```
After this transaction you should call `os.utime()` on the file to update the atime and mtime. Simply reading is likely not going to update..  `os.utime` on one file for each read should be fine -- you also should not need to do a write lock to call it on the file. It just updates metadata.

This transaction has the default timeout of 120s, which IMO is way too long. I would explicitly code something like 3s here. If you are waiting, you're waiting for another process to open and write this file, which shouldn't take much time. if it does, it means the system is just slow (maybe lots of IO?) and you should get on with concretizing and bail early by returning `None, None`

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -607,228 +643,198 @@ class ConcretizationCache:
+            # more recently, i.e. not a good candidate for LRU, or it's already being removed
+            # so we dont care. Could also be a write lock from another clean operation
+            # in which case that operation can remove it
+            with self.write_transaction(entry_to_rm, timeout=10) as exists:
```
10s is too much. Make this more `1e-6` or something really short. If you cannot get the write lock on the thing, you do not care and there is no reason to wait. It's a cache -- if you're wrong and the system is being slow right now, meh, no harm done. Spend time cleaning up again later when it's fast. If you can't get the lock for some legit reason, as you say, it's either being read or being removed -- both cases mean you should just skip removing it and move on.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -607,228 +643,198 @@ class ConcretizationCache:
-        self._manifest_queue: List[Tuple[pathlib.Path, int]] = []
+        self.root.mkdir(parents=True, exist_ok=True)
+        self._lockfile = self.root / ".cc_lock"
+        self._default_lock_timeout = 120
```
this seems long, and I would just avoid a default timeout. I think each call should really think through why it has the timeout it does.

johnwparent commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1046,17 +1076,171 @@ def __init__(self, cores=True):
+                lines.extend(strip_asp_problem(f.readlines()))
+
+        key = "\n".join(lines)
+        with open("/Users/gamblin2/full-problem.lp", "w", encoding="utf-8") as f:
```
@tgamblin was the intent to remove this? AFAICT we don't access the file again.

tgamblin replied:
yeah this should go away :) oops.

johnwparent committed:
ConcretizationCache: reduce performance overhead

Cleanup performed a good deal of unnecessary operations.
Remove the size limit, which removes the need to compute stats about the files before deciding to prune

Implement more carefully considered lock timeouts and refactor conc cache lock API to be compatible with floats

Handle lock timeouts in transaction context managers

update mod time on cache hit

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
concretizationcache: update comments

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
concretizercache: remove size_limit from schema

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
concretization cache test: remove test for bytes limit as limit is removed

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
asp.py remove trailing newlines

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Update tests for refactor changes

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
concretizationcache: enforce absolute path use in read transaction

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Docs: concretization cache path is in misc cache

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
concretization cache: cleanup cleanup docstring

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Missed a use of test

Signed-off-by: John Parent <john.parent@kitware.com>

tgamblin committed:
simplify cleanup logic and make testing more robust

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
make sure the test dir is empty before starting

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

johnwparent committed:
concretization cache: use session scoped directory generation for consistent fixture root

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
No need for readlock

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
undo new session conc cache root fixture

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Move conc cache from singleton to class instance

Signed-off-by: John Parent <john.parent@kitware.com>

johnwparent committed:
Adapt tests to non singleton conc cache

Signed-off-by: John Parent <john.parent@kitware.com>

tgamblin committed:
make mypy happy about clingo control type

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
remove some unnecessary code

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
add a test for canonicalization

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -607,228 +643,187 @@ class ConcretizationCache:
+
+        """
+        # path must be absolute at this point
+        assert path.is_absolute()
```
I tried to bump the Spack commit in `spack-packages`, which is currently at 501dae3fec28f57a0493a81054fad5a76f308a82

It turns out this PR breaks all Windows builds in pipelines because of this assertion (and probably the corresponding one in `read_transaction`). [This](https://gitlab.spack.io/spack/spack-packages/-/jobs/18564282) is an example of a failing build on Windows:
```
System.Management.Automation.RemoteException
  File "C:\builds\spack\spack-packages\.ci\tmp\spack\lib\spack\spack\bootstrap\core.py", line 426, in ensure_executables_in_path_or_raise
    if current_bootstrapper.try_search_path(executables, abstract_spec):
  File "C:\builds\spack\spack-packages\.ci\tmp\spack\lib\spack\spack\bootstrap\core.py", line 318, in try_search_path
    concrete_spec = spack.concretize.concretize_one(abstract_spec_str)
  File "C:\builds\spack\spack-packages\.ci\tmp\spack\lib\spack\spack\concretize.py", line 201, in concretize_one
    result = Solver().solve([spec], tests=tests, allow_deprecated=allow_deprecated)
  File "C:\builds\spack\spack-packages\.ci\tmp\spack\lib\spack\spack\solver\asp.py", line 4132, in solve
    result, _, _ = self.solve_with_stats(specs, **kwargs)
  File "C:\builds\spack\spack-packages\.ci\tmp\spack\lib\spack\spack\solver\asp.py", line 4120, in solve_with_stats
    result = self.driver.solve(
  File "C:\builds\spack\spack-packages\.ci\tmp\spack\lib\spack\spack\solver\asp.py", line 1253, in solve
    self._conc_cache.store(cache_key, result, self.control.statistics)
  File "C:\builds\spack\spack-packages\.ci\tmp\spack\lib\spack\spack\solver\asp.py", line 785, in store
    with self.write_transaction(cache_path, timeout=30):
  File "C:\builds\spack\spack-packages\.ci\tmp\spack\lib\spack\spack\solver\asp.py", line 773, in write_transaction
    assert path.is_absolute()
```
Is it necessary to check that the path is absolute here (I didn't read the code yet)? If it is, we should work out why in pipelines we don't have an absolute path, and add a better error message than just a blank assertion error.

For completeness the PR bumping the Spack commit is: https://github.com/spack/spack-packages/pull/989

johnwparent replied:
I don't think it needs to be absolute, we pass `path.name` to the lock anyway. I think the difference in behavior is coming from `cannonicalize_path` although not sure why that wouldn't have shown up in unit tests, so perhaps a CI setting issue.

I'll PR a quick fix for the abs paths and then a follow up to make the behavior consistent.

johnwparent replied:
See: https://github.com/spack/spack/pull/51465

johnwparent replied:
Can't replicate locally, smells like a CI config issue
